File: to-do-app/To-Do-Java/src/main/java/com/renan/todo/service/TaskService.java
package com.renan.todo.service;


import com.renan.todo.dto.TaskDTO;

import java.util.List;

/**
 * Services for tasks operation
 */
public interface TaskService {

    /**
     * Get the tasks for the list passed
     *
     * @param listID - list id
     * @param userID - user id
     *
     * @return - list of tasks for the list id and user id
     * @throws BusinessException - for invalid user, list or other errors
     */
    public List<TaskDTO> getTasksByList(Integer listID, Integer userID) throws BusinessException;

    /**
     * Delete all tasks from the list
     *
     * @param listID - list id
     * @param userID - user id
     *
     * @throws BusinessException - for invalid user, list or other errors
     */
    public void deleteTasksFromList(Integer listID, Integer userID) throws BusinessException;

    /**
     * Mark as complete all tasks from the list
     *
     * @param listID - list id
     * @param userID - user id
     *
     * @throws BusinessException - for invalid user, list or other errors
     */
    public void completeTasksFromList(Integer listID, Integer userID) throws BusinessException;

    /**
     * Mark as incomplete all tasks from the list
     *
     * @param listID - list id
     * @param userID - user id
     *
     * @throws BusinessException - for invalid user, list or other errors
     */
    public void incompleteTasksFromList(Integer listID, Integer userID) throws BusinessException;

    /**
     * Change the listID on all tasks from a list
     *
     * @param listIDOrigin  - list id from where the tasks will be updated
     * @param listIDDestiny - list id to set to the tasks
     * @param userID        - user id
     *
     * @throws BusinessException - for invalid input, list or other errors
     */
    public void moveTasksFromList(Integer listIDOrigin,
                                  Integer listIDDestiny, Integer userID) throws BusinessException;

    /**
     * Create a new task for the user
     *
     * @param taskName        - task name
     * @param taskDescription - description
     * @param deadline        - deadline
     * @param listID          - list id (the list where will be saved the new task)
     * @param userID          - user id
     *
     * @return - the new task created
     * @throws BusinessException - for invalid input, list or other errors
     */
    public TaskDTO createTask(String taskName, String taskDescription, String deadline
            , Integer listID, Integer userID) throws BusinessException;

    /**
     * Update a task on the database
     *
     * @param taskName        - task name
     * @param taskDescription - description
     * @param deadline        - deadline
     * @param taskID          - task id
     * @param userID          - user id
     *
     * @throws BusinessException - for invalid input, list or other errors
     */
    public void updateTask(String taskName, String taskDescription, String deadline
            , Integer taskID, Integer userID) throws BusinessException;

    /**
     * change the list for the task
     *
     * @param taskID        - task id
     * @param listIDDestiny - new list for the task
     * @param userID        - user id
     *
     * @throws BusinessException - for invalid input, list or other errors
     */
    public void moveTaskToList(Integer taskID, Integer listIDDestiny, Integer userID) throws BusinessException;

    /**
     * Delete the task on the database
     *
     * @param taskID - task id
     * @param userID - user id
     *
     * @throws BusinessException - for invalid input, list or other errors
     */
    public void deleteTask(Integer taskID, Integer userID) throws BusinessException;

    /**
     * Mark the task as complete
     *
     * @param taskID - task id
     * @param userID - user id
     *
     * @throws BusinessException - for invalid input, list or other errors
     */
    public void completeTask(Integer taskID, Integer userID) throws BusinessException;

    /**
     * Mark the task as incomplete
     *
     * @param taskID - task id
     * @param userID - user id
     *
     * @throws BusinessException - for invalid input, list or other errors
     */
    public void incompleteTask(Integer taskID, Integer userID) throws BusinessException;

}

File: to-do-app/to-do-kotlin/src/main/kotlin/com/renan/todo/service/TaskService.kt
package com.renan.todo.service

import com.renan.todo.dto.TaskDTO

/**
 * Services for tasks operation
 */
interface TaskService {

    /**
     * Get the tasks for the list passed
     */
    @Throws(BusinessException::class)
    fun getTasksByList(listID: Int?, userID: Int?): List<TaskDTO?>?

    /**
     * Delete all tasks from the list
     */
    @Throws(BusinessException::class)
    fun deleteTasksFromList(listID: Int?, userID: Int?)

    /**
     * Mark as complete all tasks from the list
     */
    @Throws(BusinessException::class)
    fun completeTasksFromList(listID: Int?, userID: Int?)

    /**
     * Mark as incomplete all tasks from the list
     */
    @Throws(BusinessException::class)
    fun incompleteTasksFromList(listID: Int?, userID: Int?)

    /**
     * Change the listID on all tasks from a list
     */
    @Throws(BusinessException::class)
    fun moveTasksFromList(
        listIDOrigin: Int?,
        listIDDestiny: Int?, userID: Int?
    )

    /**
     * Create a new task for the user
     */
    @Throws(BusinessException::class)
    fun createTask(
        taskName: String?, taskDescription: String?, deadline: String?, listID: Int?, userID: Int?
    ): TaskDTO?

    /**
     * Update a task on the database
     */
    @Throws(BusinessException::class)
    fun updateTask(
        taskName: String?, taskDescription: String?, deadline: String?, taskID: Int?, userID: Int?
    )

    /**
     * change the list for the task
     */
    @Throws(BusinessException::class)
    fun moveTaskToList(taskID: Int?, listIDDestiny: Int?, userID: Int?)

    /**
     * Delete the task on the database
     */
    @Throws(BusinessException::class)
    fun deleteTask(taskID: Int?, userID: Int?)

    /**
     * Mark the task as complete
     */
    @Throws(BusinessException::class)
    fun completeTask(taskID: Int?, userID: Int?)

    /**
     * Mark the task as incomplete
     */
    @Throws(BusinessException::class)
    fun incompleteTask(taskID: Int?, userID: Int?)

}